generator/internal/task_processor.go

```diff
@@ -137,6 +145,10 @@ func (p *taskProcessor) completeTask(logger lager.Logger, container executor.Con
 			p.failTask(logger, container.Guid, TaskCompletionReasonInvalidTransition)
 		}
 		return
+	} else if container.RunResult.Failed {
+		// When we have done BBS.CompleteTask on a failed Task increment the counter
+		// It will be incremented on every call to FailTask in the p.failTask() method
+		TasksFailed.Increment()
```

Comment:
Ah, we understand now the full logic. Even though we read the comment, it wasn't immediately clear that the reason for doing this extra TasksFailed.Increment() was needed in this particular spot in order to avoid duplication with the TasksFailed.Increment() in `p.failTask()`

For what is essentially pretty simple logic, it took us a lot longer to understand exactly what what is going on than it should have, and I think the basic problem with the way this is implemented is that it breaks the priniciple of having a consistent level of abstraction per function depth. If the metric is being incremented at function depth X, I don't also expect it to be incremented at function depth X+1, and that's where our surprise came from.

While we understand the merits of implementing this on the rep, we think this could be much simpler if it were implemented in the BBS with a cell-id tag, like @ematpl mentioned. Then the logic shows up in just a couple places, and the level of abstraction in the code is consistent. So, basically:

- In the Task start handler (or whatever the equivalent is), increment the TasksStarted counter
- In the CompleteTask handler, check the RunResult.Failed property. If true, increment TasksFailed, if false, increment TasksSucceeded.
- In the FailTask handler, increment the TasksFailed counter.

One detail to watch out for: Not sure off the top of my head whether a failed task also goes through the CompleteTask handler, so you'll want to double check this.